dockertest/output.py

```diff
@@ -737,11 +737,32 @@ def mustfail(cmdresult, failmsg=None):
         details = "%s\n%s" % (failmsg, cmdresult)
     if cmdresult is not None:
         OutputNotBad(cmdresult)
-    if cmdresult.exit_status == 0:
-        raise xceptions.DockerExecError("Unexpected zero exit code,"
-                                        " details: %s" % details)
+    # FIXME: temporary; remove once we no longer run on pre-1.10 docker
+    if not has_distinct_exit_codes():
+        expected_status = 1
+    if cmdresult.exit_status != expected_status:
+        raise xceptions.DockerExecError("Unexpected exit code %d; expected %d."
+                                        " Details: %s" % (cmdresult.exit_status,
+                                                          expected_status,
+                                                          details))
     return cmdresult
 
+HAS_DISTINCT_EXIT_CODES = None
+def has_distinct_exit_codes():
+    """
+    2016-03-23 **TEMPORARY** - for transition from docker-1.9 to 1.10
+
+    docker-1.10 will introduce distinct exit codes to allow differentiating
+    between container status and errors from docker run itself; see bz1097344
+    and docker PR14012. If we see an exit code of 125 here, assume we're
+    using the new docker.
+    """
+    global HAS_DISTINCT_EXIT_CODES
+    if HAS_DISTINCT_EXIT_CODES is None:
+        d_run = utils.run('docker run --invalid-option nonexistent-container',
+                          ignore_status=True)
+        HAS_DISTINCT_EXIT_CODES = (d_run.exit_status > 120)
+    return HAS_DISTINCT_EXIT_CODES
 
 # This class inherits a LOT of public methods, but most of what
 # appears here is all 'behind the scenes' stuff for producing
```

Comment:
A few ideas here:
    - Make it an immutable bool-subclass, maybe singleton-pattern.  Might be overcomplicating it though.
    - Have it accept a 'state' parameter that defaults to None, then is returned as something which evaluates true/false but is mutable (e.g. ``[True]`` or ``[False]``).
    - Have ``mustfail()`` accept a callable parameter, which defaults to ``has_distinct_exit_codes``, then it just calls it instead of looking up a global.  (You'll need to flip around the function definition order).